### PR TITLE
#464 #465 add assignments logic

### PR DESCRIPTION
**Pull request by andreievg**
Requires back end  #464F-#465F-add-self-assignment-option-on-templatePermission
Fixes: #464
Fixes: #465

changes in this PR will make assignment and self assignment work (not re-assignment or actionables in application list)

### Changes

(not targeted)

* update error checking for async mutations (noticed we were doing some extra bits inside the component, outsourced to hooks

(targeted)

* added mutation and helper hook, `useUpdateReviewAssignmet`
* changes `getReviewInfoQuery` to also get assigners through join and `templateSectionRestrictions`
* `useGetReview` hook adds, `iscurrentUserReviewer`, isCurrentUserAssigner` and `assibnableSectionRestrictions`
*  add `AssignmentSectionRow`
* changed `generateReviewSectionActions` to only return `canSelfAssign ` option if reviewer is current user (would there be a need to show all users that can self assign in review home /)

(notes)

* maybe we should do all of the calculation inside the hook vs in `AssignmentSectionRow` (but it would require combining multiple assignments into one structure)

### Testing

* using back end branch above
* create new ReviewTest application and submit
* reviewer options (assigner options, self assignment etc) should comply with the digram below
* One thing to note, self assignment and manual assignment are not meant to be working on the same application/stage/level (it's either one or the other). In this PR the only interchangeable thing is that assigner will see self assigned user and wont' be able to assign another one. 
* Self when one user self assigns, the other should't see the option anymore (via action)

Screen Shot 2021-03-25 at 12 37 42 PM

### CODE REVIEW

**Comment by CarlosNZ** [comment]
Couple of initial comments from testing:

- I can see the "Self-Assign" button for the two Sections. When I click one of them, they both get assigned to me. Is this intended behaviour?
- After assigning, when I click "Start", it just gives an error and review not started (after a refresh) -- looks like this:
Screen Shot 2021-03-25 at 2 02 16 PM

**Comment by andreievg** [comment]
> I can see the "Self-Assign" button for the two Sections. When I click one of them, they both get assigned to me. Is this intended behaviour?

Yep - I think we need a modal to say all sections will be assigned? Or only self assign that section ?

> After assigning, when I click "Start", it just gives an error and review not started (after a refresh) -- looks like this:

I am not getting this error, hmm, can you give a replication example (which back end branch, is it new application or  an application from insertData) ?

File: src/containers/Review/ReviewHome.tsx [hunk 1]
```diff
@@ -28,13 +29,17 @@ const ReviewHome: React.FC<ReviewHomeProps> = ({ assignments, structure }) => {
 
   const [filters, setFilters] = useState<Filters | null>(null)
 
+  const getFilteredByStage = (assignments: AssignmentDetailsNEW[]) => {
+    if (!filters) return []
```
Comment (CarlosNZ) [inline src/containers/Review/ReviewHome.tsx:33]: Don't you think that no filters should return ALL, not NONE?

**Review by CarlosNZ — APPROVED** [review]
All looks pretty solid and tests okay (not getting that weird but any more, don't really know why).

Will probably notice other bugs once it's properly integrated with more functionality, but seems good to merge now. Just one small question, nothing major.

**Comment by nmadruga** [comment]
Having a look on this one today. I'm willing to merge if all good in test so we can remove some of the old PRs from the list of PRs to review - It's growing to big and starting to get messy to test without other things in place.

File: src/utils/hooks/useUpdateReviewAssignment.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+    const createReviewQuestionAssignments = assignableElements.map((element) => ({
+      templateElementId: element.element.id,
+    }))
+
+    return {
+      status: ReviewAssignmentStatus.Assigned,
+      // onReviewSelfAssign will trigger core action to disallow other reviewers to self assign
+      trigger: isSelfAssignment ? Trigger.OnReviewSelfAssign : Trigger.OnReviewAssign,
+      // TODO: change to timeUpdated
```
Comment (nmadruga) [inline src/utils/hooks/useUpdateReviewAssignment.tsx:57]: Adding a placeholder issue for that

**Review by nmadruga — APPROVED** [review]
Looks pretty solid to me after testing for a little while. 
I have made just a few commits:
- Remove console.log 
- Rename UN_ASSIGN to UNASSIGN constant
- Remove `templateElementId` from queries (and replaced where used with `templateElement.id`)

Hope that's all good with you @andreievg. I'm merging this one and will create separated issues for things that might need a fix/improvement as part of ReviewAssignment Epic.

File: src/utils/hooks/useUpdateReviewAssignment.tsx (hunk 2 above)
Comment (nmadruga) [reply src/utils/hooks/useUpdateReviewAssignment.tsx:57]: Created a placeholder note for this issue [here](https://github.com/orgs/openmsupply/projects/6#card-58402523)

**Pull request merged**